### PR TITLE
NickAkhmetov/CAT-677 Unified Prov Graph 

### PR DESCRIPTION
**Pull request by NickAkhmetov**
This PR 
- enhances the provenance view by 
  - finding the top-level raw datasets the current dataset is descended from
  - finding those datasets' bottom-level descendants
  - combining those descendants' provenance trees
  - This approach allows us to view the full provenance for any dataset.
- converts the provenance graph components to TS

Currently, this functionality is gated to datasets only.

<details><summary>Examples</summary>
<p>

[HBM878.HRPG.642](https://portal.hubmapconsortium.org/browse/dataset/d8f851efb54f84d7ee0952e10d4c449e)
image

[HBM749.SMWP.555](https://portal.hubmapconsortium.org/browse/dataset/6df2f796ad72307d04dc94d688b725c5)
image

</p>
</details>

### CODE REVIEW

**Comment by john-conroy** [comment]
Will review shortly, but should we make the base branch a unified datasets feature branch?

**Review by john-conroy — COMMENTED** [review]
Should we remove the 'Show Derived Entities' button for dataset nodes?

File: context/app/static/js/components/detailPage/provenance/ProvGraph/DetailPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import { InternalLink } from 'js/shared-styles/Links';
+import ShowDerivedEntitiesButton from 'js/components/detailPage/provenance/ShowDerivedEntitiesButton';
+import { useTrackEntityPageEvent } from 'js/components/detailPage/useTrackEntityPageEvent';
+import { StyledPaper, StyledTypography } from './style';
+import '@hms-dbmi-bgm/react-workflow-viz/dist/react-workflow-viz.min.css';
+import { ProvData } from '../types';
+
+interface DetailPanelProps {
+  prov: Record<string, string>;
+  // eslint-disable-next-line react/no-unused-prop-types
```
[the remark this answers is not in the document]
Comment (NickAkhmetov) [reply context/app/static/js/components/detailPage/provenance/ProvGraph/DetailPanel.tsx:14]: For some reason this error kept popping up for me despite using every prop listed. I kept trying different workarounds for a while before settling on the eslint disable.

File: context/app/static/js/components/detailPage/provenance/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  const descendants = await fetchSearchData<{ uuid: string }, unknown>(
+    getDescendantsQuery(uuidsToFetch),
+    elasticsearchEndpoint,
+    groupsToken,
+  );
+
+  const descendantIds = descendants.hits.hits.map((hit) => hit._id);
+
+  const descendantProvenance = await Promise.all(
```
[the remark this answers is not in the document]
Comment (NickAkhmetov) [reply context/app/static/js/components/detailPage/provenance/hooks.ts:63]: I'll give it a shot, likely interpolating it into this function - my motivation for having this pulled-out `getCombinedProvData` fetcher is to make sure the combinedProvenance logic only runs once after data is fetched.

File: context/app/static/js/components/detailPage/provenance/utils.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  return `${entityEndpoint}/entities/${uuid}/provenance`;
+}
+
+/**
+ * Combines ProvData objects without overwriting properties or inserting unnecessary duplicates.
+ * @param a the first ProvData object
+ * @param b the second ProvData object
+ * @param keyMod the string to append to the key of a property in b if it already exists in a
+ */
+export function nonDestructiveMerge<T extends Record<string, unknown>, U extends Record<string, unknown>>(
```
[the remark this answers is not in the document]
Comment (NickAkhmetov) [reply context/app/static/js/components/detailPage/provenance/utils.ts:17]: The `used` and `wasGeneratedBy` fields have conflicting keys, so a traditional deep merge does not work; e.g., the sibling provenances in my screenshots have this collision in the `used` field:
```
// prov1.used:
        "_:id3": {
            "prov:activity": "hubmap:activities/02127e1a2d6211f3b5a8f63a605a5ae7",
            "prov:entity": "hubmap:entities/6df2f796ad72307d04dc94d688b725c5"
        }
// prov2.used: 
        "_:id3": {
            "prov:activity": "hubmap:activities/ff569428ddfd6c6ebd66ca652495f74b",
            "prov:entity": "hubmap:entities/6df2f796ad72307d04dc94d688b725c5"
        },
```
These two keys have the same entity, but refer to different activities, so one of the connections gets clobbered when doing a deep merge unless the key mod is applied.

**Comment by NickAkhmetov** [comment]
> should we make the base branch a unified datasets feature branch?

My understanding was that this is a change we can roll out on all dataset pages, including the legacy views.

> Should we remove the show derived entities button for dataset nodes

This question came up for me as well, I figured that it wouldn't hurt to keep it in for consistency with the donor/sample provenance graphs.

**Pull request merged**